File: API/src/main/java/com/insane/apiwtb/model/Product.java
package com.insane.apiwtb.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Builder
@Table(name = "TB_PRODUCT")
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_product")
    private int id;

    @Column(name = "nm_product", nullable = false, length = 50)
    private String name;

    @Column(name = "nr_price", precision = 9, scale = 2)
    private Double price;

    @Column(name = "ds_description", length = 200)
    private String description;

    @Column(name = "nr_discount", precision = 3)
    private Integer discount;

    @ManyToOne
    @JoinColumn(name = "id_shop")
    private Shop shop;

    @ManyToOne
    @JoinColumn(name = "id_product_type")
    private ProductType productType;

    @ManyToMany
    @JoinTable(name = "tb_product_bra_type", joinColumns = @JoinColumn(name = "id_product"), inverseJoinColumns = @JoinColumn(name = "id_bra_type"))
    private List<BraType> braTypes  = new ArrayList<>();

    @ManyToMany
    @JoinTable(name = "tb_product_category", joinColumns = @JoinColumn(name = "id_product"), inverseJoinColumns = @JoinColumn(name = "id_category"))
    private List<Category> categories  = new ArrayList<>();

    @ManyToMany
    @JoinTable(name = "tb_product_image", joinColumns = @JoinColumn(name = "id_product"), inverseJoinColumns = @JoinColumn(name = "id_image"))
    private List<Image> images = new ArrayList<>();

}

File: Mobile/Android/app/src/main/java/com/insane/whatsthebra/model/Product.kt
package com.insane.whatsthebra.model

import com.insane.whatsthebra.database.dto.ProductDTO

data class Product (
    var id: Int = 0,
    val name: String,
    val price: Double,
    val description: String,
    var discount: Int = 0,
    val shop: Shop,
    val productType: ProductType,
    val braTypes: ArrayList<BraType> = ArrayList(),
    val categories: ArrayList<Category> = ArrayList(),
    val images: ArrayList<Image> = ArrayList()) {

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + (name?.hashCode() ?: 0)
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (this === other) return true
        if (other !is Product) return false
        if (name != other.name || id != other.id) return false
        return true
    }

    fun toProductDTO(): ProductDTO {
        return ProductDTO (
            id = id,
            name = name,
            price = price,
            description = description,
            discount = discount,
            shopId = shop.id,
            productTypeId = productType.id
        )
    }

}